File: red/src/workflow.ts
import { readPars, parName } from "red/cli";
import * as dryRun from "red/dry-run";
import { preflight } from "red/lifecycle";
import * as progress from "red/progress";
import * as tofu from "red/tofu";
import { adviceAdd, failed, workflow, type Opts, type WireDecl } from "red/workflow";
import * as compute from "./compute.ts";
import * as sshConfig from "./ssh-config.ts";
import * as tools from "./tools.ts";
import * as validate from "./validate.ts";

export const defaults: Opts = {
  "provider-compute": "vultr", "provider-dns": "cloudflare",
  "provider-backend": "r2", "compute-prevent-destroy": true,
  workdir: ".colors",
};

export async function startStep(
  opts: Opts,
  env: Record<string, string | undefined> = process.env,
): Promise<Opts> {
  return preflight(opts, {
    defaults,
    overlay: readPars,
    validators: [
      (_opts, environment) => validate.envErrors(environment),
      (current) => validate.stateErrors(current),
      (current, _environment, { event, real }) =>
        real && (event === "create" || event === "delete")
          ? validate.secretErrors(current, event)
          : [],
      (current, _environment, { event, real }) =>
        real && event === "delete" && current["compute-prevent-destroy"]
          ? [`compute destruction is protected; set ${parName("compute-prevent-destroy")}=false to delete`]
          : [],
    ],
    afterValidate: (current, _environment, {event,real}) => real && event==='create' ? sshConfig.preflight({...current,'red/exit':0}) : {...current,'red/exit':0},
  }, env);
}

export function wireFn(step: string, runOpts: Opts): WireDecl | undefined {
  if (runOpts["red/event"] === "delete") {
    const graph: Record<string, WireDecl> = {
      "netbird/start": [startStep, "netbird/load"], "netbird/load": [compute.loadStep, "netbird/ansible"],
      "netbird/ansible": [tools.ansibleStep, "netbird/dns"],
      // The `~/.ssh/config` block goes before the destroy, the opposite of the
      // keypair below. A block that outlives its host is stale but harmless; a
      // key that predeceases its host locks the operator out of a machine that
      // still exists. Both orders are deliberate; see standards/ssh-config.md.
      "netbird/dns": [tools.dnsStep, "netbird/ssh-config"],
      "netbird/ssh-config": [tools.ansibleLocalStep, "netbird/infrastructure"],
      "netbird/infrastructure": [tools.infrastructureStep],
    };
    return graph[step];
  }
  const graph: Record<string, WireDecl> = {
    "netbird/start": [startStep, "netbird/infrastructure"],
    // After compute, which is where the address first exists, and before the
    // stage that converges the machine.
    "netbird/infrastructure": [tools.infrastructureStep, "netbird/ssh-config"],
    "netbird/ssh-config": [tools.ansibleLocalStep, "netbird/dns"],
    // DNS before convergence: Traefik asks Let's Encrypt for a certificate
    // the moment it starts, and TLS-ALPN-01 only succeeds once the names
    // resolve to this host. The record existing is necessary but not
    // sufficient — the playbook additionally waits for public resolvers to
    // carry it before starting anything.
    "netbird/dns": [tools.dnsStep, "netbird/ansible"],
    "netbird/ansible": [tools.ansibleStep, "netbird/acceptance"],
    "netbird/acceptance": [tools.acceptanceStep],
  };
  return graph[step];
}

export function backendAdvice(tool: string) {
  return tofu.conventionalBackendAdvice({
    dir: (opts) => tools.toolDir(opts, tool),
    key: (opts) => `${opts.profile ?? ""}/${tool}.tfstate`,
  });
}

export const sideEffecting = [
  "netbird/infrastructure", "netbird/dns", "netbird/ssh-config",
  "netbird/ansible", "netbird/acceptance", "netbird/load",
];

function create() {
  let wf = workflow({ start: "netbird/start", wireFn });
  wf = adviceAdd(wf,"netbird/dns","before","netbird.workflow/backend",backendAdvice(tools.dnsTool));
  return dryRun.advise(progress.advise(wf), sideEffecting);
}

export const netbirdWorkflow = create();

File: green/src/clj/io/github/getcolors/netbird/workflow.clj
(ns io.github.getcolors.netbird.workflow
  (:require [clojure.walk :as walk]
            [green.cli :as green-cli]
            [green.dry-run :as dry-run]
            [green.lifecycle :as lifecycle]
            [green.progress :as progress]
            [green.tofu :as tofu]
            [green.workflow :as wf]
            [io.github.getcolors.netbird.compute :as compute]
            [io.github.getcolors.netbird.ssh-config :as ssh-config]
            [io.github.getcolors.netbird.tools :as tools]
            [io.github.getcolors.netbird.validate :as validate]))

(def defaults {:provider-compute "vultr" :provider-dns "cloudflare"
               :provider-backend "r2" :compute-prevent-destroy true
               :workdir ".colors"})

(defn start-step
  ([opts] (start-step opts (System/getenv)))
  ([opts env]
   (lifecycle/preflight
    opts {:defaults defaults :overlay green-cli/read-pars
          :validators
          [(fn [_ env _] (validate/env-errors env))
           (fn [opts _ _] (validate/state-errors opts))
           (fn [opts _ {:keys [event real?]}]
             (when (and real? (contains? #{:create :delete} event))
               (validate/secret-errors opts event)))
           (fn [opts _ {:keys [event real?]}]
             (when (and real? (= :delete event) (:compute-prevent-destroy opts))
               [(str "compute destruction is protected; set "
                     (green-cli/par-name :compute-prevent-destroy) "=false to delete")]))]
          :after-validate
          (fn [opts _ {:keys [event real?]}]
            (if (and real? (= :create event)) (ssh-config/preflight! (assoc opts :green/exit 0)) (assoc opts :green/exit 0)))} env)))

(defn wire-fn [step run-opts]
  (if (= :delete (:green/event run-opts))
    (case step
      :netbird/start [start-step :netbird/load]
      :netbird/load [compute/load-step :netbird/ansible]
      :netbird/ansible [tools/ansible-step :netbird/dns]
      ;; The `~/.ssh/config` block goes before the destroy, the opposite of the
      ;; keypair below. A block that outlives its host is stale but harmless; a
      ;; key that predeceases its host locks the operator out of a machine that
      ;; still exists. Both orders are deliberate; see standards/ssh-config.md.
      :netbird/dns [tools/dns-step :netbird/ssh-config]
      :netbird/ssh-config [tools/ansible-local-step :netbird/infrastructure]
      :netbird/infrastructure [tools/infrastructure-step])
    (case step
      :netbird/start [start-step :netbird/infrastructure]
      ;; After compute, which is where the address first exists, and before the
      ;; stage that converges the machine.
      :netbird/infrastructure [tools/infrastructure-step :netbird/ssh-config]
      :netbird/ssh-config [tools/ansible-local-step :netbird/dns]
      ;; DNS before convergence: Traefik asks Let's Encrypt for a certificate
      ;; the moment it starts, and TLS-ALPN-01 only succeeds once the names
      ;; resolve to this host. The record existing is necessary but not
      ;; sufficient — the playbook additionally waits for public resolvers to
      ;; carry it before starting anything.
      :netbird/dns [tools/dns-step :netbird/ansible]
      :netbird/ansible [tools/ansible-step :netbird/acceptance]
      :netbird/acceptance [tools/acceptance-step])))

(defn backend-advice [tool]
  (tofu/conventional-backend-advice
   {:dir-fn #(tools/tool-dir % tool)
    :key-fn #(str (:profile %) "/" tool ".tfstate")}))

(def side-effecting
  [:netbird/infrastructure :netbird/dns :netbird/ssh-config
   :netbird/ansible :netbird/acceptance :netbird/load])

(def workflow
  (-> (wf/workflow {:start :netbird/start :wire-fn wire-fn})
      (wf/advice-add :netbird/dns :before ::backend (backend-advice tools/dns-tool))
      progress/advise
      (dry-run/advise side-effecting)))
